### PR TITLE
Block large payloads inside `DocumentDeltaConnection`, size due to the 1MB Kafka limit

### PR DESCRIPTION
**Pull request by andre4i**
See https://github.com/microsoft/FluidFramework/issues/7599

It would block message transmission and close the container but only with explicit feature gate.

### CODE REVIEW

**Comment by andre4i** [comment]
@markfields @vladsud please take a look at this. There is a bit of a behavior change with regards to retrying on error in the deltamanager.

File: packages/loader/container-loader/src/connectionManager.ts [hunk 1]
```diff
@@ -78,7 +78,11 @@ function getNackReconnectInfo(nackContent: INackContent) {
 const createReconnectError = (fluidErrorCode: string, err: any) =>
     wrapError(
         err,
-        (errorMessage: string) => new GenericNetworkError(fluidErrorCode, errorMessage, true /* canRetry */),
+        (errorMessage: string) => new GenericNetworkError(
+            fluidErrorCode,
+            errorMessage,
+            err?.canRetry === true || err?.canRetry === undefined, // unless explicitly specified, this will retry
```
Comment (andre4i) [inline packages/loader/container-loader/src/connectionManager.ts:84]: ^^ We would always retry, regardless of the type of error. Not sure if this has always been intentional or not..

Comment (vladsud) [reply packages/loader/container-loader/src/connectionManager.ts:84]: Here is the story :)
In ODSP, we definitely want to reconnect always, because we may get disconnect with 403 due to token expiration. 403 is critical error (i.e., in general it's a game over event), but across layers we always do one retry with refreshed token to ensure host has a chance to provide new token.

With that said, I think that's the wrong layer to participate in this game. I.e. ODSP has to project such errors as recoverable. I'd need to look at the code to say if it's the case. And not sure about FRS.

Also, worth noting that for long period of time we consider errors without canRetry to be recoverable. We changed that (maybe 6 months back) - any exception, whether it's a bug in our code, or in host code (like token callback) is catastrophic. 

So, to summarize, this code is likely wrong. But I do not think you are making it any more right :)

**Comment by vladsud** [comment]
I know you looked into it before, but I want to poke more.
Is there really no way to determine that socket.io disconnect happened due to size violation?
Is there anything we can do to observe lower-level traffic to figure it out? That would be obviously much better direction to convert such errors to catastrophic errors (not sure about 0.9Mb limit).

**Comment by andre4i** [comment]
> Is there really no way to determine that socket.io disconnect happened due to size violation?

None that I can find :( https://github.com/microsoft/FluidFramework/issues/8179#issuecomment-964626592 was the farthest I've gotten investigating/debugging this particular issue. I think the best way is to actually add the limitation to the server (our server) based on my notes here: https://github.com/microsoft/FluidFramework/issues/7599#issuecomment-961294411, push that value down using the client config, the documentdeltaconnection should read that config and either block OR if we fix the 1MB kafka limit do the fix in the runtime layer to support larger batches.

The current PR is to be explicit about a limit that is 'de facto' breaking things silently.

File: packages/loader/container-loader/src/connectionManager.ts (hunk 1 above)
Comment (andre4i) [reply packages/loader/container-loader/src/connectionManager.ts:84]: I don't believe there is a viable way right now for propagating non-retriable errors from the driver to the container. Maybe a TTL on the error itself? What are your thoughts?

File: packages/drivers/driver-base/src/messageSizeValidator.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        this.trackPayload(payloadSizeInBytes);
+        return allMessagesUnderLimit && payloadSizeInBytes < this.maxPayloadSizeInBytes;
+    }
+
+    public static sizeInBytes(message: IDocumentMessage): number {
+        const { contents, ...restOfObject } = message;
+        // `contents` is already stringified. Re-stringifying the whole message will
+        // lead to additional escape characters which will increase the size artificially.
```
Comment (vladsud) [inline packages/drivers/driver-base/src/messageSizeValidator.ts:71]: I think this comment suggest that what we measure is not what actually gets counted by socket.io. I.e. if socket.io strigifies payload, then it will add all these escape characters and they will go against the limit, right?
